build: macOS toolchain update

fanquake opened the pull request:
TLDR: This updates our macOS toolchain to use a newer version of Clang, cctools (including new [dependency on libtapi](https://github.com/tpoechtrager/cctools-port/tree/master#dependencies)), LD64 and the macOS SDK.

I've been testing depends builds (`HOST=x86_64-apple-darwin16`) inside a Debian Buster [Docker container](https://github.com/fanquake/core-review/blob/master/docker/debian.dockerfile), and running the resultant `bitcoind` and `bitcoin-qt` binaries on a macOS `10.14.4` system. The `.dmg` generated by a `make deploy` also mounts correctly on the same macOS system.

#### Clang
Upgraded from `3.7.1` to [`6.0.1`](https://releases.llvm.org/6.0.1/docs/ReleaseNotes.html)

#### cctools
* cctools `877.8` -> [`921`](https://opensource.apple.com/tarballs/cctools/)
* LD64 `253.9` -> [`409.12`](https://opensource.apple.com/source/ld64/)
* TAPI [`1000.10.8`](https://opensource.apple.com/tarballs/tapi/)

See [tpoechtrager/cctools-port](https://github.com/tpoechtrager/cctools-port/) and [tpoechtrager/apple-libtapi](https://github.com/tpoechtrager/apple-libtapi/).

#### macOS SDK
Upgraded from building against the macOS `10.11` SDK to the macOS `10.14` SDK.

#### TODO
- [x] Make the `10.14` SDK available to Travis.

Fixes: #16052
Closes: #14797

fanquake committed:
depends: clang 6.0.1

This also removes some now-unnecessary cctools hacks.

Co-Authored-By: Cory Fields <cory-nospam-@coryfields.com>

fanquake committed:
build: use macOS 10.14 SDK

Co-Authored-By: Carl Dong <accounts@carldong.me>

fanquake committed:
depends: native_cctools 921, ld64 409.12, libtapi 1000.10.8

This also removes the obsolete mlinker-version option

Co-Authored-By: Cory Fields <cory-nospam-@coryfields.com>

dongcarl commented on depends/packages/native_libdmg-hfsplus.mk:
```diff
@@ -1,8 +1,8 @@
-$(package)_file_name=libdmg-hfsplus-v$($(package)_version).tar.gz
-$(package)_sha256_hash=6569a02eb31c2827080d7d59001869ea14484c281efab0ae7f2b86af5c3120b3
+$(package)_version=3e5fd3fb56bc9ff03beb535979e33dcf83fe1f70
+$(package)_download_path=https://github.com/fanquake/libdmg-hfsplus/archive
```
Any reason why we're using your fork instead of upstream?

fanquake replied:
I'm still applying one commit on top of upstream: https://github.com/fanquake/libdmg-hfsplus/commit/3e5fd3fb56bc9ff03beb535979e33dcf83fe1f70

Although now that there's only one commit left we could put the patch in depends and point back to upstream if you'd prefer?

theuni reviewed (commented):
Nice! Concept ack.

+1 for a new prog to replace our dmg hacks. I think that might also mean that we could get rid of CMake.

Will review in depth.

practicalswift commented:
Concept ACK

Sjors commented:
Tested building depends on macOS 10.14.6 Mojave. I still had to do this annoying workaround:

```
open /Library/Developer/CommandLineTools/Packages/macOS_SDK_headers_for_macOS_10.14.pkg
```

And building protobuf fails: [log](https://gist.github.com/Sjors/2b39b5f70016b059170f65ee87f56dfb) 

If I remove protobuf and native_protobuf it does build. I can then compile QT:

```sh
./autogen.sh
./configure --prefix=`pwd`/depends/x86_64-apple-darwin18.7.0 --disable-bip70
make
```

This works, but I get a flood of warnings at the end: [log](https://gist.github.com/Sjors/2b39b5f70016b059170f65ee87f56dfb#file-warnings-log)

Will test on macOS 10.15 Catalina later.

We should backport this to 0.18 given #16032 and ideally do a 0.18.2 tag before Catalina is out (September or October). Our 0.19 release is scheduled for November #15940, so otherwise there's no tag to build from.

Concept ACK on bumping macOS from 10.10 to 10.12, though a small percentage of users would have to stay on the 0.18 branch (the last tag before backport).

The bottleneck for upgrading macOS is hardware, and that's pretty durable. Here are the minimum specs for 10.10:
<img width="871" alt="system 10 10" src="https://user-images.githubusercontent.com/10217/61854056-a01fc780-aebd-11e9-8e25-76af45077cf8.png">

Minimum specs for 10.12:
<img width="576" alt="system 10 12" src="https://user-images.githubusercontent.com/10217/61854067-a3b34e80-aebd-11e9-8f8e-936a0cecb080.png">

Some refurbished machines for sale:
<img width="480" alt="refurbished" src="https://user-images.githubusercontent.com/10217/61854110-b29a0100-aebd-11e9-9dff-a49f882d018b.png">

Some second hand machines at an even lower price point still work:
<img width="621" alt="macbook pro 2011" src="https://user-images.githubusercontent.com/10217/61854440-600d1480-aebe-11e9-9d84-37aa84dfc2e0.png">

Some don't:
<img width="626" alt="macbook pro 2009" src="https://user-images.githubusercontent.com/10217/61854428-584d7000-aebe-11e9-8495-f26884bd86e1.png">

fanquake commented:
> I still had to do this annoying workaround:

Yep, that isn't new with this change, but may go away if we can remove any reliance on having the headers in the now deprecated location. See details in the [`Command Line Tools section`](https://developer.apple.com/documentation/xcode_release_notes/xcode_10_release_notes) of the Xcode 10 release notes. 

> If I remove protobuf and native_protobuf it does build. I can then compile QT:

Did you try to build depends before installing the new headers? If you `make clean` in depends, then try rebuilding as usual it should work. Nothing Protobuf related is being changed in this PR.

jonasschnelli commented:
Concept ACK.
Unsure about the timing, if its already the time to drop 10.10 & 10.11 since 10.11 was released on Oct 2015.
According to statista, in April 2019, 10.10 and 10.11 makes about 16% of mac installations.
(https://www.statista.com/statistics/944559/worldwide-macos-version-market-share/)

Aidium publishes their update stats and it looks like, >10% was using 10.11 this week.
https://www.adium.im/sparkle/

It'd say we wait with this.

Sjors commented:
> Did you try to build depends before installing the new headers? If you `make clean` in depends

I thought I did, but maybe not. I just tried `make clean` and now protobuf indeed just buids (on 
536034b).

fanquake commented:
Updated to add some documentation (Linux) for the macOS 10.14 SDK extraction. This doesn't have to be the final docs, but should be useful for someone on a Linux machine wanting to extract the SDK and test this PR.

> It'd say we wait with this.

We don't have to bump the minimum to `10.12` right now. We could do `10.11`, or potentially leave it at `10.10` assuming there aren't any issues using the new SDK. Regardless, I'd like the rest of the changes (or at least the new Clang) to make it into `v0.19.0`, so that building for macOS isn't broken on Debian Buster #16052.

Sjors commented:
From Apple: "All Macintosh computers that can run Mountain Lion, Mavericks, or Yosemite can run El Capitan" (10.11). I suspect many people just don't realize they can upgrade.

I managed to build `depends` on macOS 10.14.6 (Mojave) with the minimum version set to 10.10

> [...] may go away if we can remove any reliance on having the headers in the now deprecated location

That means this PR is not enough to support Catalina, at least not without some other workaround.

theuni commented on depends/hosts/darwin.mk:
```diff
@@ -1,7 +1,7 @@
 OSX_SDK_VERSION=10.14
 OSX_SDK=$(SDK_PATH)/MacOSX$(OSX_SDK_VERSION).sdk
-LD64_VERSION=253.9
+LD64_VERSION=409.12
```
All of the LD64_VERSION stuff (used by darwin_CC and darwin_CXX) stuff may be unnecessary now. I believe "-mlinker-version" was a hack around a bug that may have now been fixed.

Could you try removing the ```-mlinker-version=...``` and seeing if the build still succeeds?

fanquake replied:
Have combined your LD64 suggestions into the cctools commit.

theuni commented on depends/packages/native_cctools.mk:
```diff
@@ -1,26 +1,41 @@
   $(build_SHA256SUM) -c $($(package)_extract_dir)/.$($(package)_file_name).hash && \
+  mkdir -p libtapi installed_libtapi && \
+  tar --no-same-owner --strip-components=1 -C libtapi -xf $($(package)_source_dir)/$($(package)_libtapi_file_name) && \
+  INSTALLPREFIX=$($(package)_extract_dir)/installed_libtapi ./libtapi/build.sh && \
```
The main reason for adding this dependency here rather than separately would be so that we can use our fresh compiler (clang) to build it. The same way we handle cctools.

Unless I'm misunderstanding, the way this is done currently, the system compiler is used. We've headed off a few frustrating incompatibilities by using a fixed compiler for the related toolchain components (cctools) in the past, and I think we should continue that.

Can you move the configure/build around to make that happen? Maybe handle both in _preprocess?

fanquake replied:
I've switched this around so that `libtapi` is now being built with our 6.0.1 Clang.

theuni commented on depends/packages/native_cctools.mk:
```diff
@@ -1,26 +1,41 @@
+  INSTALLPREFIX=$($(package)_extract_dir)/installed_libtapi ./libtapi/install.sh && \
   mkdir -p toolchain/bin toolchain/lib/clang/6.0.1/include && \
   tar --no-same-owner --strip-components=1 -C toolchain -xf $($(package)_source_dir)/$($(package)_clang_file_name) && \
   rm -f toolchain/lib/libc++abi.so* && \
```
Sadly I can't remember why the libc++abi.so removal is needed, maybe the SDK ships with the proper one?

As for dsymutil, that hack was added because the previous cctools didn't ship with one, and some build scripts expect to find it. If the new cctools ships with one, or if whatever required it has been fixed, this can be removed.

fanquake replied:
Combined your suggestions into the Clang update commit.

theuni commented on depends/packages/native_cctools.mk:
```diff
@@ -31,14 +46,14 @@ define $(package)_extract_cmds
 
 define $(package)_preprocess_cmds
-  cd $($(package)_build_subdir); ./autogen.sh && \
+  cd $($(package)_build_subdir); autoreconf -if && ./autogen.sh && \
```
Isn't this redundant?

fanquake replied:
Have dropped this entire line.

theuni commented on depends/packages/native_cctools.mk:
```diff
@@ -52,6 +67,9 @@ endef
   $(MAKE) DESTDIR=$($(package)_staging_dir) install && \
+  mkdir -p $($(package)_staging_prefix_dir)/lib/ && \
+  cd $($(package)_extract_dir)/installed_libtapi && \
+  cp lib/libtapi.so.6 $($(package)_staging_prefix_dir)/lib/ && \
```
We want this built statically if possible, otherwise this introduces a (rather unnecessary) LDPATH runtime requirement. Does build.sh have static lib support?

theuni reviewed (commented):
For the most part this looks great. I've added a few notes around libtapi which still needs a little work IMO.

I've not tried building yet, I'll work on getting an SDK extracted.

theuni commented:
I tested a few of my cleanup suggestions above, and they seem to work fine.
Feel free to pull the top two from here: https://github.com/theuni/bitcoin/commits/16392, or adapt them into your commits however's easiest.

fanquake commented:
@theuni I've incorporated some of your suggestions (trimming the LD64 stuff, and the clang hacks) and added you as co-author to the relevant commits.

We are now pointing back to upstream `libdmg-hfsplus`, with a single patch included in depends.

We are also now pointing back to the upstream `cctools-port` repo.

`libtapi` is now being built with our `6.0.1` Clang and I've removed one of the intermediate directories I was creating. 

I still need to address:
> We want this built statically if possible, otherwise this introduces a (rather unnecessary) LDPATH runtime requirement. Does build.sh have static lib support?

Sjors commented:
Building depends on macOS 10.14.6 Mojave still works as of 97b3349.

fanquake commented:
I've updated this to pull the macOS DMG notes (with some minor changes) out of `build-osx.md` and put them in the `contrib/macdeploy/README.md` with the other deployment notes.
I've also added a new section on extracting the new SDK when your using `macOS`.

BlockMechanic commented:
:+1:  Concept ACK, this may help alliviate issues i've had trying to create build bitcoin-qt on aarch64.

laanwj commented:
I think it's too late to do this for 0.19, might want to bump the milestone to 0.20 and then do it early in the merge window?

jonasschnelli commented:
I also think it is too early for not good reason.
> According to statista, in April 2019, 10.10 and 10.11 makes about 16% of mac installations.

fanquake commented:
> I think it's too late to do this for 0.19, 

Ok. I've punted this to `v0.20.0`, for an early merge. Will address remaining comments here shortly.

laanwj commented:
> Ok. I've punted this to v0.20.0, for an early merge. Will address remaining comments here shortly.

:smiling_imp:

dongcarl commented:
Gitian build for 329af88c548ff4afe6fed7fdbad3d39f62044048 is failing for me, `tail -n1024 build.log`: https://paste.sr.ht/%7Edongcarl/fcafb82a86a68e86c5b29b5e32e00d68211d2661

fanquake commented:
@dongcarl Looks like the macOS SDK isn't available?

dongcarl commented:
@fanquake `docker exec -it --user root gitian-target find -name '*.sdk.tar.gz'` outputs `./build/MacOSX10.14.sdk.tar.gz`

fanquake commented:
@dongcarl Has it been unzipped inside `depends/SDKs`?

Every time I've seen `fatal error: 'unistd.h' file not found` has been because depends can't find the macOS SDK.

dongcarl commented:
@fanquake Here's what I see inside the docker container:
```
root@694ebefb6b63:/home/ubuntu/build/bitcoin/depends/SDKs/MacOSX.sdk# ls
Entitlements.plist  Library  SDKSettings.json  SDKSettings.plist  System  usr
```

dongcarl commented:
Oh wait... I have too new of an Xcode... It's actually `MacOSX10.15.sdk` that I `tarred` up...

fanquake commented:
> Oh wait... I have too new of an Xcode... It's actually MacOSX10.15.sdk that I tarred up...

@dongcarl That would likely actually work fine.

The real problem is that you need to rename the directory to `MacOSX10.14.sdk` rather than `MacOSX.sdk`. See how the path is [built in depends](https://github.com/bitcoin/bitcoin/blob/master/depends/hosts/darwin.mk#L3).

dongcarl commented:
@fanquake Thanks for offline help for the above.
I also ran into another error mentioning `libc++abi.so` that was solved by readding a previously-removed line:
```diff
diff --git a/depends/packages/native_cctools.mk b/depends/packages/native_cctools.mk
index 2371a691b..ecfc6f9d5 100644
--- a/depends/packages/native_cctools.mk
+++ b/depends/packages/native_cctools.mk
@@ -36,6 +36,7 @@ define $(package)_extract_cmds
   tar --no-same-owner --strip-components=1 -C libtapi -xf $($(package)_source_dir)/$($(package)_libtapi_file_name) && \
   mkdir -p toolchain/bin toolchain/lib/clang/6.0.1/include && \
   tar --no-same-owner --strip-components=1 -C toolchain -xf $($(package)_source_dir)/$($(package)_clang_file_name) && \
+  rm -f toolchain/lib/libc++abi.so* && \
   tar --no-same-owner --strip-components=1 -xf $($(package)_source)
 endef
```

dongcarl commented:
With these patches, I was able to perform a Gitian build: https://github.com/fanquake/bitcoin/compare/329af88c548ff4afe6fed7fdbad3d39f62044048...dongcarl:2019-10-fixing-fanquake-osx-bump

BlockMechanic commented:
I do not see an actual change to Qt 5.12 in the PR... is this in preparation ?

Fuzzbawls commented:
Regarding the bumping of minimum supported macOS (OS X) version: I would say that 10.11 minimum is fine, but 10.12 requires newer CPU based SSE4.1 support that prevent otherwise perfectly viable CPUs (older Xeon CPUs mainly) from upgrading to 10.12.

MarcoFalke commented on contrib/macdeploy/README.md:
````diff
@@ -1,15 +1,130 @@
+```
 
-This script should not be run manually, instead, after building as usual:
+During the deployment process, the disk image window will pop up briefly 
````
```suggestion
During the deployment process, the disk image window will pop up briefly
```

:eyes:

fanquake commented:
@dongcarl pulled your changes in as part of 1052e1a44e93a3a3cf2fe18a17916ea5f2dbe07f.

dongcarl commented:
@fanquake Also, as mentioned above, you probably also need f6448f99ef1cb688edffcd4b5934af938b0e228a to fix Gitian builds. :grimacing:

fanquake commented:
@dongcarl I've reverted the removal of that line in bf9488b1dd228a011f61ae21a642ee9ce69e03c2 as well updated 9d51eb8cca106ecaf9a05e52e15bfbf93fbde4a8 to point to the latest upstream version of `cctools`.

dongcarl commented:
@fanquake Would you say that this is no longer WIP? I'd say it's ready for merge :grimacing:

MarcoFalke commented:
Not so fast. Let's fix the travis and DrahtBot failure first. ;)

MarcoFalke commented:
https://travis-ci.org/bitcoin/bitcoin/jobs/598728043#L1466

fanquake commented:
> Would you say that this is no longer WIP? I'd say it's ready for merge 😬

I'll do a final rebase/any cleanup and remove the `[WIP]` tag.

> Not so fast. Let's fix the travis and DrahtBot failure first. ;)

This should just require making the newer SDK available to both.

MarcoFalke commented on .travis.yml:
```diff
@@ -157,6 +157,6 @@ jobs:
 
     - stage: test
-      name: 'macOS 10.10  [GOAL: deploy] [no functional tests]'
+      name: 'macOS 10.12  [GOAL: deploy] [no functional tests]'
```
So the macOS version is `10.12`, the SDK version is `10.14`, and the cross compile target is `x86_64-apple-darwin16`? This is not confusing at all.

fanquake replied:
`x86_64-apple-darwin16` maps to macOS `10.12`, more info [here](https://en.wikipedia.org/wiki/Darwin_%28operating_system%29#Release_history). We're building with the 10.14 SDK, however this can be newer, and is the same as we've been doing in previous releases.

MarcoFalke commented on contrib/gitian-build.py:
```diff
@@ -208,7 +208,7 @@ def main():
 
     # Disable for MacOS if no SDK found
-    if args.macos and not os.path.isfile('gitian-builder/inputs/MacOSX10.11.sdk.tar.gz'):
+    if args.macos and not os.path.isfile('gitian-builder/inputs/MacOSX10.14.sdk.tar.gz'):
```
Should the gitian-build in master support compiling older releases?

dongcarl replied:
Not sure that's a burden we wanna take on. It's probably going to be more sane for those with that use-case to check out the older tag.

MarcoFalke replied:
Seems fine

MarcoFalke commented on doc/release-notes.md:
```diff
@@ -46,15 +46,14 @@ Compatibility
-From Bitcoin Core 0.17.0 onwards, macOS versions earlier than 10.10 are no
-longer supported, as Bitcoin Core is now built using Qt 5.9.x which requires
-macOS 10.10+. Additionally, Bitcoin Core does not yet change appearance when
+From 0.19.0 onwards, macOS <10.12 is no longer supported.
```
```suggestion
From 0.20.0 onwards, macOS <10.12 is no longer supported.
```

?

Sjors commented on contrib/macdeploy/README.md:
```diff
@@ -1,15 +1,135 @@
-### MacDeploy ###
+# Mac Deployment
```
Nit: macOS; more SEO and search friendly

Sjors commented on contrib/macdeploy/README.md:
````diff
@@ -1,15 +1,135 @@
+```
+
+Our previously used macOS SDK (`MacOSX10.11.sdk`) can be extracted from
+[Xcode 7.3.1 dmg](https://developer.apple.com/devcenter/download.action?path=/Developer_Tools/Xcode_7.3.1/Xcode_7.3.1.dmg).
````
Maybe update this URL and move it to the top of this section: https://download.developer.apple.com/Developer_Tools/Xcode_10.3/Xcode_10.3.xip (from July 2019)

In general this section refers to several different versions of Xcode. For SDK 10.14 we need Xcode 10. A more future proof instruction might be :`This can be found here: https://developer.apple.com/download/more/ by filtering "xcode 10"`

fanquake replied:
Why update this URL, it's meant to be pointing to the previously used SDK? The link to the newer SDK is later in the file.

Sjors replied:
Ah ok, let's just drop that altogether then (can be done in a followup, along with bumping xcode 10.2.1 to 10.3 and putting the URL above the place where you need it instead of below).

Sjors commented on contrib/macdeploy/README.md:
```diff
@@ -1,15 +1,135 @@
+## SDK Extraction
 
-When finished, it will produce `Bitcoin-Qt.dmg`.
+After Xcode version 7.x, Apple started shipping the `Xcode.app` in a `.xip` archive.
```
Nit: we can lose the historical reference and just say "`Xcode.app` is packaged in a `.xip` archive.

Sjors reviewed (commented):
Tested 3525067 on macOS 10.15 Catalina. Although QT depends are built and configure log says `checking whether to build Bitcoin Core GUI... yes (Qt5)`, the configure summary drops it: `with gui / qt = no`. See configure [log](https://gist.github.com/Sjors/aaf0114a9eb13fd822556e5d806dceef).

* `extract-osx-sdk.sh` needs to be updated as (though I've never used it)
* macOS deploy instructions still refer to 10.11
* there's a reference to `MAC_OS_X_VERSION_MIN_REQUIRED > 101100` in `optionsdialog.cpp` which presumably can be dropped?

fanquake commented on depends/hosts/darwin.mk:
```diff
@@ -1,4 +1,4 @@
-OSX_MIN_VERSION=10.10
+OSX_MIN_VERSION=10.12
 OSX_SDK_VERSION=10.11
```
Looks like you're commenting on an outdated diff. This is `10.14` in 1728e348c7678504502fc28db109cc870e286695.

fanquake commented:
Rebased on master and pushed some changes to fix the doc suggestions above.

> Although QT depends are built and configure log says checking whether to build Bitcoin Core GUI... yes (Qt5), the configure summary drops it: with gui / qt = no. See configure log.

That's unrelated to this PR. The reason Qt isn't being picked up is in the log: `configure: WARNING: PROTOC not found; bitcoin-qt frontend will not be built`.

> there's a reference to MAC_OS_X_VERSION_MIN_REQUIRED > 101100 in optionsdialog.cpp which presumably can be dropped?

There are multiple similar changes that can be made post this PR, but I'm not going to do them here.

Sjors commented:
Re: `WARNING: PROTOC not found`: that means it's not finding protobuf in the depends. I'm also getting this issue on master, so indeed unrelated.

fanquake commented:
Me and Cory have been discussing moving straight to Clang 9, instead of 6. However that has turned up some issues related to Boost and Clangs `-dumpversion`. Cory is going to PR some changes first, then we can circle back around to here.

fanquake commented:
Pushed up a rebase here as reqeusted by theuni. Going to look at using the 10.15 SDK instead of the 10.14 SDK as well.

MarcoFalke commented:
> Pushed up a rebase here as reqeusted by theuni. Going to look at using the 10.15 SDK instead of the 10.14 SDK as well.

The 10.15 SDK gave problems, so sticking to 10.14 seems better for now? https://github.com/bitcoin-core/secp256k1/issues/674#issuecomment-546520166

MarcoFalke commented:
Anything left to do here except pushing up the new SDK to our depends mirror?

fanquake commented:
> Anything left to do here except pushing up the new SDK to our depends mirror?

I have some cctools-port changes to review and we will likely switch to a newer version of Clang. I will get that done over the next 2 days.

fanquake commented:
Sorry for the delay. I've pushed up the discussed changes. This is now based on top of #17550. I've also reordered the commits to have the SDK bump (10.14) at the end. 

One thing left TODO is diff the the 10.14 & 10.15 headers. As well as the LLVM headers. Still going to do this even though we can't use the 10.15 SDK (Qt 5.9 or 5.12 wont build with it).

MarcoFalke commented:
Idk why DrahtBot doesn't compile this. I am pretty sure I updated it to the newest sdk.

Empact commented:
FYI the latest 10.11 security update from Apple was published in 09 Jul 2018.
https://support.apple.com/en-us/HT201222

Sjors commented:
tACK 248526e4f941edea8851a5a72a2a1da0ffdfabcf on macOS 10.15.3 (also tried `make deploy`), provided you can make Travis happy too. I'm not a build system connaisseur though, so an ACK from @theuni would be great.

We should drop the instructions for obtaining the 10.11 SDK, but that can wait for a followup.

laanwj commented:
FWIW https://bitcoincore.org/depends-sources/sdks/MacOSX10.14.sdk.tar.gz is up

jonasschnelli commented:
Tested ACK via a depends build with macOS SDK 10.14:
https://bitcoinbuilds.org/?job=e969d43f-2cd8-4890-b5ae-bcc2d0ea4195

hebasto commented:
> FWIW https://bitcoincore.org/depends-sources/sdks/MacOSX10.14.sdk.tar.gz is up

Downloaded file
```
$ sha256sum MacOSX10.14.sdk.tar.gz 
2530c1d88f1aca98994cf32cfcf6099d880965ef0200b8b2a87279c191d56766  MacOSX10.14.sdk.tar.gz
```
seems broken.

hebasto commented:
Tested ca5055a5aa07aba81a87cf12f6f0526a63c423b5

Cross compilation on Linux Mint 19.3 with depends and MacOSX10.14.sdk was contaminated by system Qt packages (#18042, #18045, #18051):
```
$ make HOST=x86_64-apple-darwin16 -C depends
$ ./autogen.sh
$ CONFIG_SITE=$PWD/depends/x86_64-apple-darwin16/share/config.site ./configure
$ make
...
  CXX      qt/qt_libbitcoinqt_a-peertablemodel.o
qt/notificator.cpp:50:21: error: allocation of incomplete type 'QDBusInterface'
    interface = new QDBusInterface("org.freedesktop.Notifications",
                    ^~~~~~~~~~~~~~
...
```

The patch from #18051 was applied, then compilation passed successfully with loads of linker warnings like this:
```
  AR       qt/libbitcoinqt.a
  CXXLD    qt/bitcoin-qt
  CXXLD    qt/test/test_bitcoin-qt
ld: warning: direct access in function '__ZN5boost10filesystem6detail9canonicalERKNS0_4pathES4_PNS_6system10error_codeE' from file '/home/hebasto/GitHub/bitcoin/depends/x86_64-apple-darwin16/lib/libboost_filesystem-mt-x64.a(operations.o)' to global weak symbol '__ZZN5boost6system16generic_categoryEvE25generic_category_instance' from file 'libbitcoin_server.a(libbitcoin_server_a-init.o)' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
...
```

The resulted `Bitcoin-Core.dmg` was successfully installed on macOS 10.14 then launched .

fanquake commented:
@MarcoFalke Looks like DrahtBot is still missing the new SDK.

Sjors commented:
re-ACK 7e2104433cd0905ccf94632511b3ca0ce5b0463b (rebased from 248526e)

dongcarl commented:
@fanquake I apologize for my lateness, but here's my outstanding changes so that we can generate and use the SDK with headers as requested: https://github.com/fanquake/bitcoin/compare/macos-toolchain-update...dongcarl:2020-01-macos-sdk-with-headers

Here are the new instructions: https://github.com/dongcarl/bitcoin/blob/adf9968de5dc872e3c51e538f6b99e9423b65bd3/contrib/macdeploy/README.md#step-2-generating-macosx1014sdktargz-from-xcodeapp

laanwj commented:
Let's try to get this merged soon—if we still want this for 0.20 it's not something that should be done last minute.

dongcarl commented:
ACK 7e21044, let's merge this.

Empact commented:
Built depends successfully on 10.14.6

Is there anything more to say re the "obsolete mlinker-version option"? Is it obsolete only in that it is no longer required for build?

laanwj commented:
> @fanquake I apologize for my lateness, but here's my outstanding changes so that we can generate and use the SDK with headers as requested: fanquake/bitcoin@macos-toolchain-update...dongcarl:2020-01-macos-sdk-with-headers

Thanks, please add these instructions in a new PR!

dongcarl commented:
> > @fanquake I apologize for my lateness, but here's my outstanding changes so that we can generate and use the SDK with headers as requested: fanquake/bitcoin@macos-toolchain-update...dongcarl:2020-01-macos-sdk-with-headers
> 
> Thanks, please add these instructions in a new PR!

It's here! :-) #18072
